src/dropboxHelper.js

```diff
@@ -0,0 +1,121 @@
+// Functions for storing notes to dropbox
+import { pathJoin, readBlobAsString } from './utils';
+import { dropboxConfig } from './userConfig';
+import { noteNameFromIdAndTitle } from './NoteData';
+
+const NOTES_FOLDER_PATH = pathJoin(
+    dropboxConfig.notesFolderParent === '/' ? '' : dropboxConfig.notesFolderParent,
+    dropboxConfig.notesFolderName
+);
+
+// Helper class for saving and retrieving notes from dropbox.
+class DropboxHelper {
+    // TODO: Remove this ctor altogether
+    constructor(notesFolderPath = NOTES_FOLDER_PATH) {
+        this.dbx = undefined;
+        this.notesFolderPath = notesFolderPath;
+    }
+
+    // Will be initialized after we complete setting up dropbox.
+    init(dbx) {
+        console.assert(!!dbx);
+        this.dbx = dbx;
+    }
+
+    isInitialized() {
+        return !!this.dbx;
+    }
+
+    // Save given editor value. We currently store only the JSON representation of the note.
+    // TODO: store plain text representation alongside it.
+    async save(noteData) {
+        if (!this.dbx) {
+            console.log('dropbox was not initialized. so not saving.');
+            return undefined;
+        }
+
+        if (!noteData.videoId || !noteData.videoTitle) {
+            console.warn('No videoId or videoTitle given');
+            return undefined;
+        }
+
+        const fileName = noteNameFromIdAndTitle(noteData.videoId, noteData.videoTitle);
+        const filePath = pathJoin(this.notesFolderPath, fileName);
+        const writeMode = { '.tag': 'overwrite' };
+
+        return this.dbx.filesUpload({
+            contents: JSON.stringify(noteData),
+            mode: writeMode,
+            path: filePath,
+            autorename: false,
+        });
+    }
+
+    async downloadAllNoteFiles() {
+        if (!this.isInitialized()) {
+            console.warn('Dropbox NOT initialized.');
+            return undefined;
+        }
+
+        // Assumes notes folder already exists
+        const listFolderResult = await this.dbx.filesListFolder({
+            path: NOTES_FOLDER_PATH,
+            recursive: true,
+        });
+
+        const noteFileNames = [];
+
+        for (let entry of listFolderResult.entries) {
+            if (entry['.tag'] !== 'file') {
+                continue;
+            }
+            noteFileNames.push(entry.name);
+        }
+
+        // Download each.
+        const downloadResultPromises = [];
+        for (const noteFileName of noteFileNames) {
+            const path = pathJoin(NOTES_FOLDER_PATH, noteFileName);
+            const p = this.dbx.filesDownload({ path });
+            downloadResultPromises.push(p);
+        }
+
+        const downloadInfos = await Promise.all(downloadResultPromises);
```

Comment:
There might be too many connections to dropbox. Dropbox might throw too many requests error. Instead either request one at a time, or throttle to few requests at time.

Reply:
Is there a way to get multiple files in a single request?

Reply:
Yes, it via filesDownloadZip - http://dropbox.github.io/dropbox-sdk-js/Dropbox.html#filesDownloadZip__anchor

Didn't do it in this attempt due to the added work of downloading and unzipping. Can go the zip file way if you want. Manual throttling would probably be simpler to do.

Reply:
Cool. Manual throttling should be enough for now.